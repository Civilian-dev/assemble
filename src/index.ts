export * from './assemble'
export * from './types'
