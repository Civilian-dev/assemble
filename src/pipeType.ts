import { Assign } from 'utility-types'
import {
  Objectify,
  PipeFunctions,
  PipeFunctionsSync,
  ReturnTypesIntersection
 } from './types'

/**
 * Apply a series of (potentially async) functions to an interface.
 * Resolves with accumulated outputs and ignores any non-object returns.
 */
// export function pipeType <
//   Props extends object,
//   Funcs extends OptionalPipeFunctionsTuple<Props> = undefined
// > (...fns: PipeFunctions<Props>) {
//   return (props: Props = {} as any) =>
//     fns.reduce(async (prev: Promise<Props> | Props, fn) => {
//       const acc = await prev
//       const cur = await fn(acc)
//       return (typeof cur === 'object')
//         ? { ...acc, ...cur }
//         : acc
//     }, props) as Promise<ExpectedPipeReturnType<Props, Funcs>>
// }

export function pipeTypeSyncReduce<
  Props extends object
> (props: Props, funcs: PipeFunctionsSync<Props>) {
  return funcs.reduce((acc, fn) => {
    const cur = fn(acc)
    return (typeof cur === 'object')
      ? { ...acc, ...cur }
      : acc
  }, props)
}

/**
 * Apply a series of functions to an interface.
 * Returns accumulated outputs and ignores any non-object returns.
 */
// export function pipeTypeSync<
//   Funcs extends PipeFunctionsSync<any>,
//   Props extends PipeFunctionsProps<Funcs>
// > (...funcs: Funcs): (props: Props) => ExpectedPipeReturnType<Props, Readonly<Funcs>> {
//   return (props: Props = {} as any) => pipeTypeSyncReduce(props, funcs)
// }

// export function pipeTypeInlineSync<
//   Props extends object
// > (funcs: PipeFunctionsSync<Props>): (props: Props) => Props {
//   return (props: Props = {} as any) => pipeTypeSyncReduce(props, funcs)
// }

export type PipeFunctionsProps<Funcs extends PipeFunctions<any>> =
  Funcs extends PipeFunctions<infer Props> ? Props : object

export function pipeTypeSync<
  Props extends PipeFunctionsProps<Funcs>,
  Funcs extends PipeFunctionsSync<any>
// >(...funcs: Funcs): (props?: Props) => PipeReturnType<Props, Funcs> {
>(...funcs: Funcs): (props?: Props) => Assign<
  Objectify<Props>,
  Objectify<ReturnTypesIntersection<Funcs>>
> {
  return (props = {} as any) => pipeTypeSyncReduce(props, funcs)
}

export function pipeTypeInlineSync<
  Props extends object
>(...funcs: PipeFunctionsSync<Props>): (props?: Props) => Props {
  return (props = {} as any) => pipeTypeSyncReduce(props, funcs)
}
