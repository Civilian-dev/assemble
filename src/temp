export function pipeTypeSync<
  Funcs extends PipeFunctionsSync<any>,
  Props extends PipeFunctionsProps<Funcs>
>(...fns: Funcs) {
  return (props: Props = {} as any) =>
    fns.reduce((acc, fn) => {
      const cur = fn(acc) as object | void
      return (typeof cur === 'object')
        ? { ...acc, ...cur }
        : acc
    }, props) as ExpectedPipeReturnType<Props, Readonly<Funcs>>
}
