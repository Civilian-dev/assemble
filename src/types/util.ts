/** Generic function signature, for shorthand utility types. */
export type FunctionLiteral = (...args: any) => unknown

/** Generic object signature, for shorthand utility types. */
export type ObjectLiteral = { [key: string]: unknown }

/** Type that can a key of given type or undefined */
export type OptionalKeyOf<T> = keyof T | undefined

/**
 * Union of picked properties by key/s where key can be omitted and type will be undefined.
 * @example
 *   type AB = { a: any, b: any }
 *   type PickA = OptionalPick<AB, 'a'> // 👈 { a: any }
 *   type PickB = OptionalPick<AB, 'b'> // 👈 { b: any }
 *   type PickMaybeB = OptionalPick<AB, 'b' | undefined> // 👈 { b: any } | undefined
 *   type PickAB = OptionalPick<AB, 'a' | 'b'> // 👈 { a: any, b: any }
 *   type PickNone = OptionalPick<AB> // 👈 undefined
 * @todo Cannot yet do partial undefined, e.g:
 *   OptionalPick<AB, ['a' | undefined, 'b']>
 *   // ☝️ { a: any, b?: any }
 */
export type OptionalPick<T, K extends OptionalKeyOf<T> = undefined> =
  K extends keyof T ? Pick<T, K> : undefined

/**
 * Pick all required keys from given type.
 * @example
 *   type ABC = { a?: any, b: any, c: any }
 *   type RequiredKeysABC = PickRequireKeys<ABC>
 *   // ☝️ { b: any, c: any }
 */
 export type PickRequireKeys<T> = Pick<T, {
  [K in keyof T]-?: {} extends Pick<T, K> ? never : K;
}[keyof T]>

/**
 * Make all optional except given keys.
 * @example
 *   type ABC = { a: any, b: any, c: any }
 *   type OnlyAB = Only<ABC, 'a' | 'b'>
 *   // ☝️ OnlyAB = { a, b, c? }
 */
export type Only<T, K extends keyof T> = Partial<T> & Required<Pick<T, K>>

/**
 * Makes given keys optional.
 * @example
 *   type ABC = { a: any, b: any, c: any }
 *   type OptionalAB = Optional<ABC, 'a' | 'b'>
 *   // ☝️ OptionalAB = { a?, b?, c }
 */
export type Optional<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>

 /**
 * Make given keys required.
 * @example
 *   type ABC = { a?: any, b?: any, c?: any }
 *   type IncludeAB = Include<ABC, 'a' | 'b'>
 *   // ☝️ IncludeAB = { a: any, b: any, c?: any }
 */
export type Include<T, K extends keyof T> = T & Required<Pick<T, K>>

/**
 * Get Promise resolve type.
 * @example
 *   type Response = PromiseType<Promise<string>>;
 *   // ☝️ Response = string
 */
export type MaybeAwaited<T> = T extends PromiseLike<infer U> ? U : T;

/**
 * Get nested type at key.
 * @example
 *   type ABC = { a: string, b: boolean, c: number }
 *   type UnpackC = Unpack<ABC, 'c'>
 *   // ☝️ UnpackC = number
 */
export type Unpack<T, K> = K extends keyof T ? T[K] : never;

/**
 * Make a tuple from unpacked types on an interface.
 * @example
 *   type ABC = { a: string, b: boolean, c: number }
 *   type SpreadAB = Spread<ABC, ['a', 'b']>
 *   // ☝️ SpreadAB = [string, boolean]
 */
export type Spread<T, K extends Array<keyof T>> = {
  [I in keyof K]: Unpack<T, K[I]>
}

/**
 * Find one type within a union type (return `never` if not found).
 * @example
 *   type A = void | string
 *   type FindVoidA = Find<A, void>
 *   // ☝️ FindVoidA = void
 *   type B = number | string
 *   type FindVoidB = Find<B, void>
 *   // ☝️ FindVoidB = never
 */
export type Find<T, X> = T extends X ? T : never

/**
 * Get Promise resolve type.
 * @example
 *   type Response = UnwrapPromise<Promise<string>>;
 *   // ☝️ Response = string
 */
export type UnwrapPromise<T> = T extends PromiseLike<infer U> ? U : T;

/**
 * Type guard for Promise type.
 * @example
 *   const logResolved = (msg: Promise<string> | string) =>
 *     isPromise(msg) ? msg.then(console.log) : console.log(msg)
 */
export function isPromise<T> (maybePromise: Promise<T> | T): maybePromise is Promise<T> {
  return (
    typeof maybePromise === "object" &&
    typeof (maybePromise as Promise<T>).then === "function"
  )
}

/**
 * Omit undefined or void types to return only defined values (or never).
 * Similar to `NonNullable<T>` but also filters on void type.
 * @example
 *   type A = Defined<{ foo: true }> // 👈 { foo: true; }
 *   type B = Defined<'foo'> // 👈 "foo"
 *   type C = Defined<true> // 👈 true
 *   type D = Defined<undefined> // 👈 never
 *   type E = Defined<null> // 👈 never
 *   type F = Defined<never> // 👈 never
 */
export type Defined<T> = [T] extends [never | void | null] ? never : T

/**
 * Create a union type from types in a tuple.
 * @example
 *   type TupleABC = readonly ['a', 'b', 'c']
 *   type UnionABC = TupleUnion<TupleABC>
 *   // ☝️ 'a' | 'b' | 'c'
 */
export type TupleUnion<T extends ReadonlyArray<unknown>> = T[number]
