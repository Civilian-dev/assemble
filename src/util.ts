/** Get intersection type given union type `U`. */
export type UnionToIntersection<U> =
  (U extends unknown ? (k: U) => void : never) extends
  (k: infer I) => void ? I : never;

/** Assign properties of types in union type to one type. */
export type MergeUnion<U> =
  (U extends unknown ? (k: U) => void: never) extends
  (k: infer I) => void ? { [P in keyof I]: I[P] } : never

/** Generic function signature for utility type constraints. */
export type UnknownFunction = (...args: never) => unknown

/** Generic object signature for utility type constraints. */
export type UnknownObject = Record<string, unknown>

/** Assertion to narrow type of unknown to an object. */
export type Objectify<T> = T extends UnknownObject ? T : never

/** Assign attributes of intersecting types to one type. */
export type MergeIntersection<T> = { [P in keyof T]: T[P] }

/** Type that can a key of given type or undefined. */
export type OptionalKeyOf<T> = keyof T | undefined | void

/**
 * Union of picked properties by key/s where key can be omitted and type will be undefined.
 * @example
 *   type AB = { a: any, b: any }
 *   type PickA = OptionalPick<AB, 'a'> // 👈 { a: any }
 *   type PickB = OptionalPick<AB, 'b'> // 👈 { b: any }
 *   type PickMaybeB = OptionalPick<AB, 'b' | undefined> // 👈 { b: any } | undefined
 *   type PickAB = OptionalPick<AB, 'a' | 'b'> // 👈 { a: any, b: any }
 *   type PickNone = OptionalPick<AB> // 👈 undefined
 * @todo Cannot yet do partial undefined, e.g:
 *   OptionalPick<AB, ['a' | undefined, 'b']>
 *   // ☝️ { a: any, b?: any }
 */
export type OptionalPick<T, K extends OptionalKeyOf<T>> =
  K extends keyof T ? Pick<T, K> : K extends void ? void : undefined

/**
 * Pick all required keys from given type.
 * @example
 *   type ABC = { a?: any, b: any, c: any }
 *   type RequiredKeysABC = PickRequireKeys<ABC>
 *   // ☝️ { b: any, c: any }
 */
 export type PickRequireKeys<T> = Pick<T, {
  [K in keyof T]-?: {} extends Pick<T, K> ? never : K;
}[keyof T]>

/**
 * Make all optional except given keys.
 * @example
 *   type ABC = { a: any, b: any, c: any }
 *   type OnlyAB = Only<ABC, 'a' | 'b'>
 *   // ☝️ OnlyAB = { a, b, c? }
 */
export type Only<T, K extends keyof T> = Partial<T> & Required<Pick<T, K>>

/**
 * Makes given keys optional.
 * @example
 *   type ABC = { a: any, b: any, c: any }
 *   type OptionalAB = Optional<ABC, 'a' | 'b'>
 *   // ☝️ OptionalAB = { a?, b?, c }
 */
export type Optional<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>

 /**
 * Make given keys required.
 * @example
 *   type ABC = { a?: any, b?: any, c?: any }
 *   type IncludeAB = Include<ABC, 'a' | 'b'>
 *   // ☝️ IncludeAB = { a: any, b: any, c?: any }
 */
export type Include<T, K extends keyof T> = T & Required<Pick<T, K>>

/**
 * Get nested type at key.
 * @example
 *   type ABC = { a: string, b: boolean, c: number }
 *   type UnpackC = Unpack<ABC, 'c'>
 *   // ☝️ UnpackC = number
 */
export type Unpack<T, K> = K extends keyof T ? T[K] : never;

/**
 * Make a tuple from unpacked types on an interface.
 * @example
 *   type ABC = { a: string, b: boolean, c: number }
 *   type SpreadAB = Spread<ABC, ['a', 'b']>
 *   // ☝️ SpreadAB = [string, boolean]
 */
export type Spread<T, K extends Array<keyof T>> = {
  [I in keyof K]: Unpack<T, K[I]>
}

/**
 * Find one type within a union type (return `never` if not found).
 * @example
 *   type A = void | string
 *   type FindVoidA = Find<A, void>
 *   // ☝️ FindVoidA = void
 *   type B = number | string
 *   type FindVoidB = Find<B, void>
 *   // ☝️ FindVoidB = never
 */
export type Find<T, X> = T extends X ? T : never

/**
 * Get Promise resolve type.
 * @example
 *   type Response = UnwrapPromise<Promise<string>>;
 *   // ☝️ Response = string
 */
export type UnwrapPromise<T> = T extends PromiseLike<infer U> ? U : T;

/**
 * Type guard for Promise type.
 * @example
 *   const logResolved = (msg: Promise<string> | string) =>
 *     isPromise(msg) ? msg.then(console.log) : console.log(msg)
 */
export function isPromise<T> (maybePromise: Promise<T> | T): maybePromise is Promise<T> {
  return (
    typeof maybePromise === "object" &&
    typeof (maybePromise as Promise<T>).then === "function"
  )
}

/**
 * Omit undefined or void types to return only defined values (or never).
 * Similar to `NonNullable<T>` but also filters on void type.
 * @example
 *   type A = Defined<{ foo: true }> // 👈 { foo: true; }
 *   type B = Defined<'foo'> // 👈 "foo"
 *   type C = Defined<true> // 👈 true
 *   type D = Defined<undefined> // 👈 never
 *   type E = Defined<null> // 👈 never
 *   type F = Defined<never> // 👈 never
 */
export type Defined<T> = [T] extends [never | void | null] ? never : undefined extends T ? never : T

/**
 * Create a union type from types in a tuple.
 * @example
 *   type ArrayABC = ['a', 'b', 'c']
 *   type UnionABC = ArrayUnion<ArrayABC>
 *   // ☝️ 'a' | 'b' | 'c'
 */
export type ArrayUnion<T extends ReadonlyArray<unknown>> = T[number]

export type MapReturnType<T> = {
  [K in keyof T]: T[K] extends UnknownFunction ? ReturnType<T[K]> : T[K]
}

export type MapUnwrapPromises<T> = {
  [K in keyof T]: UnwrapPromise<T[K]>
}

export type MapDefined<T> = {
  [K in keyof T]: Defined<T[K]>
}

/**
 * Type intersection of all unconditional return types from array of functions (unwraps promises).
 * @example
 *   type TestFunctions = [
 *     () => { a: true },
 *     () => Promise<{ b: true }>,
 *     () => { c?: true },
 *     () => void,
 *     () => undefined,
 *     () => never
 *   ]
 *   type TestReturns = ReturnTypesIntersection<TestFunctions>
 *   // ☝️ { a: true } & { b: true } & { c?: true }
 */
export type ReturnTypesIntersection<Funcs extends UnknownFunction[]> =
 UnionToIntersection<ArrayUnion<MapDefined<MapUnwrapPromises<MapReturnType<Funcs>>>>>
