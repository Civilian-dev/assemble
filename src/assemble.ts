import { MergeUnion } from './util'
import { MixedAssemblers, SyncAssemblers, AssembledProps, AssemblyProps } from './types'

/**
 * Apply a series of (potentially async) functions to an interface.
 * Resolves with intersection of input and all unconditional function returns.
 */
export function assemble<
  Funcs extends MixedAssemblers<any>, // eslint-disable-line @typescript-eslint/no-explicit-any
  Props extends AssemblyProps<Funcs>,
  Assigned extends AssembledProps<Funcs>
> (...funcs: Funcs) {
  return <
    Input extends Props,
    Returned extends Promise<MergeUnion<Props & Input & Assigned>>
  > (props: Input) =>
    funcs.reduce(async (prev: Promise<Props> | Props, fn) => {
      const acc = await prev
      const cur = await fn(acc)
      return (typeof cur === 'object')
        ? { ...acc, ...cur }
        : acc
    }, props) as unknown as Returned
}

/**
 * Compose an array of functions to operate on an interface.
 * Returns intersection of input and all unconditional function returns.
 */
export function assembleSync<
  Funcs extends SyncAssemblers<any>, // eslint-disable-line @typescript-eslint/no-explicit-any
  Props extends AssemblyProps<Funcs>,
  Assigned extends AssembledProps<Funcs>
> (...funcs: Funcs) {
  return <
    Input extends Props,
    Returned extends MergeUnion<Props & Input & Assigned>
  > (props: Input) =>
    funcs.reduce((acc: Props, fn) => {
      const cur = fn(acc)
      return (typeof cur === 'object')
        ? { ...acc, ...cur }
        : acc
    }, props) as unknown as Returned
}

/**
 * @todo Find a solution that doesn't require overriding Returned types.
 * Return types are forced by `as unknown as Returned` to avoid errors because `Returned` could be
 * instantiated with a different subtype of the MergeUnion constraint. This seems unavoidable to
 * create the union of assigned props from different functions, with those function's input types.
*/
