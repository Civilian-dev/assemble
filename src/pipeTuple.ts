import { AsyncPipeFunction, PipeFunction, PipeFunctions, PipeReturnType } from './types'

/**
 * Strictly typed pipes infer their return type by merging return types of all composed functions.
 * Where composed pipe functions provide a property, it is asserted to non-null on the return type.
 * However, there is some declaration overhead, requiring pipe functions to be given as a tuple.
 */
export function pipeTuple <Props> (
  fns: Readonly<PipeFunctions<Props>>
) {
  return (props: Props = {} as any) =>
    fns.reduce(async (prev: Promise<Props> | Props, fn) => {
      const acc = await prev
      const cur = await fn(acc)
      return (typeof cur === 'object')
        ? { ...acc, ...cur }
        : acc
    }, props) as Promise<PipeReturnType<Props, typeof fns>>
}
// as Promise<PipeReturnType<Funcs, Props>Props>



const pipeFuncs = [funcOne, funcTwo] as const


const props = pipeTuple(pipeFuncs)()

props.then(({ }) => console.log())
