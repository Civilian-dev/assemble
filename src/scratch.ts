import {
  pipeTypeInlineSync,
  pipeTypeSync
} from './pipeType'
import { PipeFunction } from './types'

interface Props { one?: boolean, two?: boolean, three?: boolean}
const funcOne: PipeFunction<Props, 'one'> = () => ({ one: true })
const funcTwo: PipeFunction<Props, 'two'> = ({ one }) => ({ two: !!one })

const inferPipeSync = pipeTypeSync(funcOne, funcTwo)

const inlinePipeSync = pipeTypeInlineSync<Props>(
  () => ({ two: true }),
  funcTwo,
  () => ({ three: true })
)

inferPipeSync().two.valueOf
inlinePipeSync().two?.valueOf

// -------------------------------------------------------------------------------------------------

// const inferPipe = pipeType(...funcs)
// const typedPipe = pipeType<Props>(...funcs)
// const tuplePipe = pipeType<Props, typeof funcs>(...funcs)

// inferPipe().then((props) => props.one?.valueOf) // props are inferred from typed functions
// typedPipe().then((props) => props.one?.valueOf) // props are given and used in inline function definitions
// tuplePipe().then((props) => props.one.valueOf) // props and function types are given and output is non-conditional

// type ArrayType<T> = T extends Array<infer U> ? U : T;
