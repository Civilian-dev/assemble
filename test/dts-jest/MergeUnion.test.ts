import test from '../testType'
import { MergeUnion } from '../../src/util'

// @dts-jest:group MergeUnion
{
  // @dts-jest:snap {foo}|{bar}=={foo,bar}
  test<MergeUnion<{ foo: true } | { bar: true }>>()
}
