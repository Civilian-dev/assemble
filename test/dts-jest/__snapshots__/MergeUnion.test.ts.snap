// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MergeUnion {foo}|{bar}=={foo,bar} (type) should match snapshot 1`] = `"{ foo: true; bar: true; }"`;
